### PR TITLE
[DDW-197] Ledger and Trezor hardware wallet support

### PR DESCRIPTION
**Pull request by tomothespian**
This PR introduces Hardware wallets support implementation.

## Todos

- https://docs.google.com/document/d/1EuGAoZw4s28SuX--JxrLt3dUpQAEHoSIoWYqGVZyQT8/edit
- [x] Add screenshots
- [x] Remove unnecessary logging
- [x] Investigate Testnet support (Testing needed: https://github.com/input-output-hk/daedalus/compare/feature/ddw-197-implement-ledger-hw-support...feature/ddw-197-implement-ledger-hw-support-with-testnet-support?expand=1)
- [x] Improve Connect your device status message: [Connect the \"{walletName}\" device](https://github.com/input-output-hk/daedalus/blob/feature/ddw-197-implement-ledger-hw-support/source/renderer/app/i18n/locales/en-US.json#L645) [@nikolaglumac] 
- [x] Fix purging Ledger HW local storage data on Daedalus start [@nikolaglumac]
- [x] Fix issue with software wallet spending password reset in the "Delegation" wizard [@nikolaglumac]
- [x] Update copy and remove the word "currently" from the instructions on the "Pairing" dialog [@nikolaglumac]
- [ ] ~Implement Ledger alternate wallet handling (always ask for exporting pub key)~ - To be implemented in [DDW-510](https://jira.iohk.io/browse/DDW-510)
- [ ] ~Implement Trezor passphrase handling~ [GH Issue](https://github.com/input-output-hk/daedalus/issues/2238) - To be implemented in [DDW-511](https://jira.iohk.io/browse/DDW-511)
- [ ] ~Implement dynamic TTL setting~ - To be implemented in [DDW-512](https://jira.iohk.io/browse/DDW-512)
- [ ] **[Known Issue]** ~Ledger unable to handle singing transactions without outputs (e.g. Delegation requests)~ - To be resolved once VacuumLabs team fixes the issue in the Ledger Cardano App [DDW-513](https://jira.iohk.io/browse/DDW-513)

### Issues
1. Disabled state for [this red checkbox](https://user-images.githubusercontent.com/2539345/100599427-141f7a80-3311-11eb-8466-873da0fa6f7d.png) is wrong. We are missing part with 30% opacity.
2. Strange storybook behaviour for the [spinner in the button](https://user-images.githubusercontent.com/2539345/100620291-4f30a680-332f-11eb-88f1-ddcce48da188.png) 

## Screenshots

**1. Pair / Connect (EN)**

Pair - 1

Screen Shot 2020-11-30 at 11 36 00

Pair - 2

_Ledger_

Pair - 3

_Trezor_

Screen Shot 2020-11-30 at 11 30 33

Pair - 4

**_Error messages_**
 
Rejected on device

Pair - error

**1. Pair / Connect (JA)**
Pair - 1 - Ja-Jp

Screen Shot 2020-11-30 at 11 35 49

Pair - 2- Ja-Jp

_Ledger_

Pair - 3- Ja-Jp

_Trezor_

Screen Shot 2020-11-30 at 11 30 50

Pair - 4- Ja-Jp

**_Error messages_**
 
Rejected on device

Pair - Error - Ja-Jp

**2.TX Send (EN)**

Tx send - 1

Tx send - 2

Tx send - 3

Tx send - 4

Tx send - 5

**_Error messages_**
Tx send - Rejected on device

Tx send - Wrong device

**2. TX Send (JA)**

Tx send - 1 Ja-Jp

Tx send - 2 - Ja-Jp

Tx send - 3 Ja-Jp

Tx send - 4 Ja-Jp

Tx send - 5 Ja-Jp

Tx send - 6 Ja-Jp

**_Error messages_**

Tx send - Rejected on device Ja-Jp

Tx send - Wrong device Ja-Jp

**3. Delegation (EN)**

Delegate - Intro 1

Delegate - Intro 2

Delegate - Intro 3

Delegate - 1

Delegate - 2

Delegate - 3

Delegate - 4

Delegate - 5

**_Error messages_**

Delegate - rejected on device

**3. Delegation (JA)**

Delegate - Intro 1 Ja jp

Delegate - Intro 2 Ja jp

Delegate - Intro 3 Ja jp

Delegate - 1 Ja jp

Delegate - 2 Ja jp

Delegate - 3 Ja jp

Delegate - 4 Ja jp

Delegate - 5 Ja jp

**_Error messages_**

Delegate - Rejected on device Ja jp

**4. Common**

**_Summary screen_**

Summary screen

Summary screen - Ja-jp

**_Delegation Center_**

Delegation center

Delegation center - Ja-Jp

**_Device checking failures_**

_a. Wrong Firmware_

Wrong firmware JA-JP

Wrong firmware

_b. Trezor Bridge not installed_

Trezor Bridge error JA-JP

Trezor Bridge error

_c. Unsupported device_

Unsupported - Ja jp

Unsupported

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)
- [ ] TBD

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/common/types/electron-store.types.js [hunk 1]
```diff
@@ -14,6 +14,8 @@ export type StorageKey =
   | 'DATA-LAYER-MIGRATION-ACCEPTANCE'
   | 'READ-NEWS'
   | 'WALLETS'
+  | 'HARDWARE-WALLETS'
+  | 'HARDWARE_WALLET_DEVICES'
```
Comment (nikolaglumac) [inline source/common/types/electron-store.types.js:18]: @tomothespian please replace `_` with `-` here 🙏

Comment (tomothespian) [reply source/common/types/electron-store.types.js]: @nikolaglumac fixed

**Comment by ManusMcCole** [comment]
Hi @tomothespian. On windows I am recieving an error "cannot read property model of type undefined. See screenshot. I will check the other  platforms now 
image

**Comment by nikolaglumac** [comment]
> Hi @tomothespian. On windows I am recieving an error "cannot read property model of type undefined. See screenshot. I will check the other platforms now
> image

@ManusMcCole please make sure there is no AdaLite running. Simplest way is to restart your machine and try again. Make sure you have Trezor Bridge installed too. Please try this and get back to us.

**Comment by ManusMcCole** [comment]
Hi @Nikola and @tomy1610 .I restarted my windows machine and I got the same result except there is an extra "getProcess"error message in the console. I have had trezor succesfully working on both my windows and mac machine yesterday so its not an issue with bridge. Here is the <a href="https://drive.google.com/file/d/1EGUSddIP-cbBcQkIwIrJOSZZJmEzoTUk/view?usp=sharing">screenshot</a>. I had

**Comment by nikolaglumac** [comment]
> screenshot

@ManusMcCole did you enter the pin? Are you sure the device is connected to the right machine?

**Comment by ManusMcCole** [comment]
> > screenshot
> 
> @ManusMcCole did you enter the pin? Are you sure the device is connected to the right machine?

Yeah pin was entered and definetly plugged in to the right machine. After about 6 attempts it worked but im not conviced its working properly as I got the same error on my mac with this build Im going to continue testing now and I will update you

**Comment by tomothespian** [comment]
> Hi @nikola and @tomothespian .I restarted my windows machine and I got the same result except there is an extra "getProcess"error message in the console. I have had trezor succesfully working on both my windows and mac machine yesterday so its not an issue with bridge. Here is the [screenshot](https://drive.google.com/file/d/1EGUSddIP-cbBcQkIwIrJOSZZJmEzoTUk/view?usp=sharing). I had

@ManusMcCole Please add more details.
1. You tried to add new wallet?
2. You plugged-in device before you started Daedalus or after?
3. Device was unlocked?
4. Did you removed Daedalus state directory or not?
5. Plese also provide Daedalus.log file

**Comment by ManusMcCole** [comment]
Hi @tomothespian and @nikolaglumac. It is working now on both mac and windows. It eventually asked me to update to firmware 2.3.4 and now that has resolved all the issues. sorry about the delay in confirming that.This message popped up in the console it didnt stop it working but I thought I would bring it to your attention just incase.

 Moving on to check Linux now

**Comment by ManusMcCole** [comment]
Hi @tomothespian. On Linux I am having difficulties connecting the trezor. This is likely caused by complications caused by the fact that it is a VM.I have the Trezor bridge and the newest firmware I can connect to Trezor.io and the device is recognised by the machine. This same issue happened on builds that Sam and Michael were able to get it working on Linux so likely the cause is the VM issues. Here is a screenshot. I will keep trying my best to figure out the problem. Here is a <a href="https://drive.google.com/file/d/13Yyu8A43MNeDAGj0-LcZtk8lUpjMUzeP/view?usp=sharing">screenshot</a>

**Comment by nikolaglumac** [comment]
> Hi @tomothespian. On Linux I am having difficulties connecting the trezor. This is likely caused by complications caused by the fact that it is a VM.I have the Trezor bridge and the newest firmware I can connect to Trezor.io and the device is recognised by the machine. This same issue happened on builds that Sam and Michael were able to get it working on Linux so likely the cause is the VM issues. Here is a screenshot. I will keep trying my best to figure out the problem. Here is a [screenshot](https://drive.google.com/file/d/13Yyu8A43MNeDAGj0-LcZtk8lUpjMUzeP/view?usp=sharing)

@ManusMcCole please ask someone to test this on a real Linux machine. We can not deal with VM issues now.

**Comment by tomothespian** [comment]
@ManusMcCole you can also try to connect on AdaLite and check if you have same problem.

**Comment by gnpf** [comment]
@tomothespian I just tested this on Ubuntu 20.04 (not VM) with a Ledger Nano S device. It was unlocked and recognized by the machine. I try connecting the device before and after starting Daedalus but got no luck.
[This error](https://drive.google.com/file/d/1PlwNl70ET50YZt-jA3jopzipAWFmn3gP/view?usp=sharing) was displayed when clicking on "connect". [Here](https://drive.google.com/file/d/1RsWa7xr5gUJiCzfGDxWrMgc2zQxkVNAH/view?usp=sharing) are the logs.

**Comment by nikolaglumac** [comment]
> @tomothespian I just tested this on Ubuntu 20.04 (not VM) with a Ledger Nano S device. It was unlocked and recognized by the machine. I try connecting the device before and after starting Daedalus but got no luck.
> [This error](https://drive.google.com/file/d/1PlwNl70ET50YZt-jA3jopzipAWFmn3gP/view?usp=sharing) was displayed when clicking on "connect". [Here](https://drive.google.com/file/d/1RsWa7xr5gUJiCzfGDxWrMgc2zQxkVNAH/view?usp=sharing) are the logs.

@gnpf please install TrezorBridge: https://wallet.trezor.io/#/bridge

File: installer-clusters.cfg [hunk 2]
```diff
@@ -1 +1 @@
-mainnet mainnet_flight testnet shelley_qa staging
```
Comment (nikolaglumac) [inline installer-clusters.cfg:1]: @tomothespian we must not forget to revert this change prior to merging this PR 🙏

**Comment by mrbrinker** [comment]
Due to pools using the same ticker, which some times is not spoofing but also in some case it is a spoofing attempt. 

Would be good in addition to the pool ticker, the pool ID to be displayed in the delegation confirmation screen. 

See image for reference to which step of the delegation process I meant. 
delegation Ledger

**Comment by nikolaglumac** [comment]
> Due to pools using the same ticker, which some times is not spoofing but also in some case it is a spoofing attempt.
> 
> Would be good in addition to the pool ticker, the pool ID to be displayed in the delegation confirmation screen.
> 
> See image for reference to which step of the delegation process I meant.
> delegation Ledger

Hey @mrbrinker 👋 
We can definitely improve that!
Will add it to our list and we'll work on it in our next sprint.
Thanks for helping out!

**Comment by nikolaglumac** [comment]
> Due to pools using the same ticker, which some times is not spoofing but also in some case it is a spoofing attempt.
> 
> Would be good in addition to the pool ticker, the pool ID to be displayed in the delegation confirmation screen.
> 
> See image for reference to which step of the delegation process I meant.
> delegation Ledger

@mrbrinker would this work? 

Screenshot 2020-12-03 at 09 53 00

**Comment by mrbrinker** [comment]
@nikolaglumac Yes, it does work very well. 
Thank you

**Review by gnpf — APPROVED** [review]

**Review by nikolaglumac — APPROVED** [review]
Titanic work @tomothespian 💯  ❤️  🙇

**Review by a-rukin — APPROVED** [review]
LGTM
